File: shared/session_bindings.py
from __future__ import annotations

import json
import os
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from tempfile import NamedTemporaryFile
from typing import Any
from uuid import uuid4

from shared.db.sqlite.engine import database_path


_DEFAULT_AGENT_NAME = "agent"
_DEFAULT_ENTRYPOINT = "main"


def _utc_now_text() -> str:
    return datetime.now(timezone.utc).isoformat()


def _clean_text(value: Any) -> str:
    return str(value or "").strip()


def _clean_extra(value: Any) -> dict[str, Any]:
    return dict(value) if isinstance(value, dict) else {}


def _normalize_chat_type(value: Any) -> str:
    text = _clean_text(value).lower()
    if text in {"p2p", "private", "dm", "direct"}:
        return "dm"
    if text in {"group", "chat"}:
        return "group"
    if text in {"channel", "thread"}:
        return text
    return text or "dm"


def session_bindings_path() -> Path:
    configured = _clean_text(os.getenv("AGENT_SESSION_BINDINGS_PATH"))
    if configured:
        return Path(configured).expanduser()
    return database_path().parent / "sessions.json"


@dataclass(slots=True)
class SessionSource:
    platform: str
    chat_id: str
    chat_type: str
    user_id: str = ""
    user_id_alt: str = ""
    user_name: str = ""
    chat_name: str = ""
    thread_id: str = ""
    message_id: str = ""
    root_id: str = ""
    parent_id: str = ""
    is_bot: bool = False
    extra: dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, value: dict[str, Any] | None) -> "SessionSource":
        raw = dict(value or {})
        return cls(
            platform=_clean_text(raw.get("platform")),
            chat_id=_clean_text(raw.get("chat_id")),
            chat_type=_normalize_chat_type(raw.get("chat_type")),
            user_id=_clean_text(raw.get("user_id")),
            user_id_alt=_clean_text(raw.get("user_id_alt")),
            user_name=_clean_text(raw.get("user_name")),
            chat_name=_clean_text(raw.get("chat_name")),
            thread_id=_clean_text(raw.get("thread_id")),
            message_id=_clean_text(raw.get("message_id")),
            root_id=_clean_text(raw.get("root_id")),
            parent_id=_clean_text(raw.get("parent_id")),
            is_bot=bool(raw.get("is_bot")),
            extra=_clean_extra(raw.get("extra")),
        )

    def normalized(self) -> "SessionSource":
        return SessionSource.from_dict(self.to_dict())

    @property
    def user_key(self) -> str:
        return _clean_text(self.user_id_alt) or _clean_text(self.user_id)

    @property
    def session_key(self) -> str:
        source = self.normalized()
        platform = _clean_text(source.platform)
        chat_type = _normalize_chat_type(source.chat_type)
        chat_id = _clean_text(source.chat_id)
        if not platform:
            raise RuntimeError("session source platform required")
        if not chat_id:
            raise RuntimeError("session source chat_id required")
        prefix = f"{_DEFAULT_AGENT_NAME}:{_DEFAULT_ENTRYPOINT}:{platform}"
        if chat_type == "dm":
            return f"{prefix}:dm:{chat_id}"
        if chat_type == "group":
            thread_id = _clean_text(source.thread_id)
            if thread_id:
                return f"{prefix}:group:{chat_id}:{thread_id}"
            user_key = source.user_key
            if not user_key:
                raise RuntimeError("group session source user_id or user_id_alt required")
            return f"{prefix}:group:{chat_id}:{user_key}"
        return f"{prefix}:{chat_type}:{chat_id}"

    def to_dict(self) -> dict[str, Any]:
        data = asdict(self)
        data["chat_type"] = _normalize_chat_type(data.get("chat_type"))
        return {
            key: value
            for key, value in data.items()
            if value is not None and value != "" and not (key == "extra" and not value)
        }


@dataclass(slots=True)
class SessionBindingEntry:
    session_key: str
    session_id: str
    created_at: str
    updated_at: str
    origin: dict[str, Any]
    platform: str
    chat_type: str
    resume_pending: bool = False
    suspended: bool = False

    @classmethod
    def from_dict(cls, value: dict[str, Any]) -> "SessionBindingEntry":
        raw = dict(value or {})
        origin = dict(raw.get("origin") or {})
        return cls(
            session_key=_clean_text(raw.get("session_key")),
            session_id=_clean_text(raw.get("session_id")),
            created_at=_clean_text(raw.get("created_at")),
            updated_at=_clean_text(raw.get("updated_at")),
            origin=origin,
            platform=_clean_text(raw.get("platform") or origin.get("platform")),
            chat_type=_normalize_chat_type(raw.get("chat_type") or origin.get("chat_type")),
            resume_pending=bool(raw.get("resume_pending")),
            suspended=bool(raw.get("suspended")),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_key": self.session_key,
            "session_id": self.session_id,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "origin": dict(self.origin or {}),
            "platform": self.platform,
            "chat_type": self.chat_type,
            "resume_pending": bool(self.resume_pending),
            "suspended": bool(self.suspended),
        }


class SessionBindingStore:
    def __init__(self, path: Path | None = None) -> None:
        self.path = Path(path) if path is not None else session_bindings_path()

    def load_all(self) -> dict[str, SessionBindingEntry]:
        if not self.path.is_file():
            return {}
        try:
            raw = json.loads(self.path.read_text(encoding="utf-8") or "{}")
        except json.JSONDecodeError as exc:
            raise RuntimeError(f"invalid sessions.json: {self.path}") from exc
        if not isinstance(raw, dict):
            raise RuntimeError(f"sessions.json must be a JSON object: {self.path}")
        entries: dict[str, SessionBindingEntry] = {}
        for key, value in raw.items():
            if not isinstance(value, dict):
                continue
            entry = SessionBindingEntry.from_dict(value)
            session_key = _clean_text(entry.session_key or key)
            if session_key and entry.session_id:
                entry.session_key = session_key
                entries[session_key] = entry
        return entries

    def save_all(self, entries: dict[str, SessionBindingEntry]) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        payload = {
            key: entry.to_dict()
            for key, entry in sorted(dict(entries or {}).items())
            if key and entry.session_id
        }
        encoded = json.dumps(payload, ensure_ascii=False, indent=2)
        with NamedTemporaryFile(
            "w",
            encoding="utf-8",
            dir=str(self.path.parent),
            delete=False,
            prefix=f".{self.path.name}.",
            suffix=".tmp",
        ) as fh:
            fh.write(encoded)
            fh.write("\n")
            temp_path = Path(fh.name)
        temp_path.replace(self.path)

    def get(self, session_key: str) -> SessionBindingEntry | None:
        return self.load_all().get(_clean_text(session_key))

    def bind(self, source: SessionSource, *, session_id: str) -> SessionBindingEntry:
        normalized = source.normalized()
        session_key = normalized.session_key
        now = _utc_now_text()
        entries = self.load_all()
        existing = entries.get(session_key)
        created_at = existing.created_at if existing else now
        entry = SessionBindingEntry(
            session_key=session_key,
            session_id=_clean_text(session_id),
            created_at=created_at,
            updated_at=now,
            origin=normalized.to_dict(),
            platform=_clean_text(normalized.platform),
            chat_type=_normalize_chat_type(normalized.chat_type),
            resume_pending=bool(existing.resume_pending) if existing else False,
            suspended=bool(existing.suspended) if existing else False,
        )
        entries[session_key] = entry
        self.save_all(entries)
        return entry

    def get_or_create(self, source: SessionSource) -> SessionBindingEntry:
        normalized = source.normalized()
        session_key = normalized.session_key
        entries = self.load_all()
        existing = entries.get(session_key)
        if existing is not None and existing.session_id:
            return existing
        return self.bind(normalized, session_id=uuid4().hex)

    def rotate(self, source: SessionSource) -> SessionBindingEntry:
        return self.bind(source.normalized(), session_id=uuid4().hex)


def build_session_source(value: dict[str, Any] | None) -> SessionSource:
    return SessionSource.from_dict(value)


__all__ = [
    "SessionBindingEntry",
    "SessionBindingStore",
    "SessionSource",
    "build_session_source",
    "session_bindings_path",
]

File: apps/gateway/src/session-bindings.ts
import {
  mkdirSync,
  readFileSync,
  renameSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { dirname, join } from "node:path";
import { randomUUID } from "node:crypto";

export type JsonRecord = Record<string, unknown>;

const clean = (value: unknown): string => String(value ?? "").trim();
const normalizeChatType = (value: unknown): string => {
  const text = clean(value).toLowerCase();
  if (["p2p", "private", "dm", "direct"].includes(text)) return "dm";
  if (["group", "chat"].includes(text)) return "group";
  if (text === "channel" || text === "thread") return text;
  return text || "dm";
};

export interface SessionSourceData {
  platform?: unknown;
  chat_id?: unknown;
  chat_type?: unknown;
  user_id?: unknown;
  user_id_alt?: unknown;
  user_name?: unknown;
  chat_name?: unknown;
  thread_id?: unknown;
  message_id?: unknown;
  root_id?: unknown;
  parent_id?: unknown;
  is_bot?: unknown;
  extra?: unknown;
}

export class SessionSource {
  private constructor(
    readonly platform: string,
    readonly chat_id: string,
    readonly chat_type: string,
    readonly user_id: string,
    readonly user_id_alt: string,
    readonly user_name: string,
    readonly chat_name: string,
    readonly thread_id: string,
    readonly message_id: string,
    readonly root_id: string,
    readonly parent_id: string,
    readonly is_bot: boolean,
    readonly extra: JsonRecord,
  ) {}

  static from(value: SessionSourceData | null | undefined): SessionSource {
    const raw = value ?? {};
    return new SessionSource(
      clean(raw.platform),
      clean(raw.chat_id),
      normalizeChatType(raw.chat_type),
      clean(raw.user_id),
      clean(raw.user_id_alt),
      clean(raw.user_name),
      clean(raw.chat_name),
      clean(raw.thread_id),
      clean(raw.message_id),
      clean(raw.root_id),
      clean(raw.parent_id),
      Boolean(raw.is_bot),
      raw.extra !== null && typeof raw.extra === "object" && !Array.isArray(raw.extra)
        ? { ...(raw.extra as JsonRecord) }
        : {},
    );
  }

  get userKey(): string {
    return this.user_id_alt || this.user_id;
  }

  get sessionKey(): string {
    if (!this.platform) throw new Error("session source platform required");
    if (!this.chat_id) throw new Error("session source chat_id required");
    const prefix = `agent:main:${this.platform}`;
    if (this.chat_type === "dm") return `${prefix}:dm:${this.chat_id}`;
    if (this.chat_type === "group") {
      if (this.thread_id) return `${prefix}:group:${this.chat_id}:${this.thread_id}`;
      if (!this.userKey) throw new Error("group session source user_id or user_id_alt required");
      return `${prefix}:group:${this.chat_id}:${this.userKey}`;
    }
    return `${prefix}:${this.chat_type}:${this.chat_id}`;
  }

  toJSON(): JsonRecord {
    const data: JsonRecord = {
      platform: this.platform,
      chat_id: this.chat_id,
      chat_type: this.chat_type,
      user_id: this.user_id,
      user_id_alt: this.user_id_alt,
      user_name: this.user_name,
      chat_name: this.chat_name,
      thread_id: this.thread_id,
      message_id: this.message_id,
      root_id: this.root_id,
      parent_id: this.parent_id,
      is_bot: this.is_bot,
      extra: { ...this.extra },
    };
    return Object.fromEntries(
      Object.entries(data).filter(
        ([key, value]) => value !== null && value !== "" && !(key === "extra" && Object.keys(value as object).length === 0),
      ),
    );
  }
}

export interface SessionBindingEntry {
  session_key: string;
  session_id: string;
  created_at: string;
  updated_at: string;
  origin: JsonRecord;
  platform: string;
  chat_type: string;
  resume_pending: boolean;
  suspended: boolean;
}

interface StoreOptions {
  now?: () => string;
  id?: () => string;
}

const parseEntry = (value: JsonRecord): SessionBindingEntry => {
  const origin =
    value.origin !== null && typeof value.origin === "object" && !Array.isArray(value.origin)
      ? { ...(value.origin as JsonRecord) }
      : {};
  return {
    session_key: clean(value.session_key),
    session_id: clean(value.session_id),
    created_at: clean(value.created_at),
    updated_at: clean(value.updated_at),
    origin,
    platform: clean(value.platform ?? origin.platform),
    chat_type: normalizeChatType(value.chat_type ?? origin.chat_type),
    resume_pending: Boolean(value.resume_pending),
    suspended: Boolean(value.suspended),
  };
};

export class SessionBindingStore {
  private readonly now: () => string;
  private readonly id: () => string;

  constructor(readonly path: string, options: StoreOptions = {}) {
    this.now = options.now ?? (() => new Date().toISOString());
    this.id = options.id ?? (() => randomUUID().replaceAll("-", ""));
  }

  loadAll(): Record<string, SessionBindingEntry> {
    let rawText: string;
    try {
      rawText = readFileSync(this.path, "utf8");
    } catch (error) {
      const code = error instanceof Error && "code" in error ? String(error.code) : "";
      if (code === "ENOENT") return {};
      throw error;
    }
    let raw: unknown;
    try {
      raw = JSON.parse(rawText || "{}");
    } catch {
      throw new Error(`invalid sessions.json: ${this.path}`);
    }
    if (raw === null || typeof raw !== "object" || Array.isArray(raw)) {
      throw new Error(`sessions.json must be a JSON object: ${this.path}`);
    }
    const entries: Record<string, SessionBindingEntry> = {};
    for (const [key, value] of Object.entries(raw)) {
      if (value === null || typeof value !== "object" || Array.isArray(value)) continue;
      const entry = parseEntry(value as JsonRecord);
      const sessionKey = entry.session_key || clean(key);
      if (sessionKey && entry.session_id) entries[sessionKey] = { ...entry, session_key: sessionKey };
    }
    return entries;
  }

  saveAll(entries: Readonly<Record<string, SessionBindingEntry>>): void {
    mkdirSync(dirname(this.path), { recursive: true });
    const sorted: Record<string, SessionBindingEntry> = {};
    for (const key of Object.keys(entries).sort()) {
      const entry = entries[key];
      if (key && entry?.session_id) sorted[key] = entry;
    }
    const temporary = join(
      dirname(this.path),
      `.${this.path.split(/[\\/]/).at(-1)}.${randomUUID().replaceAll("-", "")}.tmp`,
    );
    try {
      writeFileSync(temporary, `${JSON.stringify(sorted, null, 2)}\n`, "utf8");
      renameSync(temporary, this.path);
    } finally {
      rmSync(temporary, { force: true });
    }
  }

  get(sessionKey: string): SessionBindingEntry | undefined {
    return this.loadAll()[clean(sessionKey)];
  }

  bind(source: SessionSource, sessionId: string): SessionBindingEntry {
    const sessionKey = source.sessionKey;
    const entries = this.loadAll();
    const existing = entries[sessionKey];
    const now = this.now();
    const entry: SessionBindingEntry = {
      session_key: sessionKey,
      session_id: clean(sessionId),
      created_at: existing?.created_at || now,
      updated_at: now,
      origin: source.toJSON(),
      platform: source.platform,
      chat_type: source.chat_type,
      resume_pending: existing?.resume_pending ?? false,
      suspended: existing?.suspended ?? false,
    };
    entries[sessionKey] = entry;
    this.saveAll(entries);
    return entry;
  }

  getOrCreate(source: SessionSource): SessionBindingEntry {
    const existing = this.get(source.sessionKey);
    return existing?.session_id ? existing : this.bind(source, this.id());
  }

  rotate(source: SessionSource): SessionBindingEntry {
    return this.bind(source, this.id());
  }
}
